src/Mapper.cpp

```diff
@@ -1346,3 +1346,223 @@ map<int, CGRANode*>* Mapper::getReorderPath(map<CGRANode*, int>* t_path) {
+                for(auto tile : tiles){
+                        idx += 1;
+                        if(idx < numTiles) jsonFile<<"          \""<<tile->getID()<<"\":"<<level<<","<<endl;
+                        else jsonFile<<"             \""<<tile->getID()<<"\":"<<level<<endl;
+                }
+        }
+        jsonFile<<"     }"<<endl;
+
+        jsonFile<<"}"<<endl;
+}
+
+// read from the referenced mapping results json file and generates variables for incremental mapping
+int Mapper::readRefMapRes(CGRA* t_cgra, DFG* t_dfg){
+        ifstream refFile("./increMapInput.json");
+        if(!refFile.good()){
+                cout<<"Incremental mapping requires increMapInput.json in current directory!"<<endl;
+                return -1;
+        }
+        json refs;
+        refFile >> refs;
+        CGRANodeID2Level.clear();
+        for(list<DFGNode*>::iterator dfgNode=t_dfg->nodes.begin(); dfgNode != t_dfg->nodes.end(); ++dfgNode){
+                int dfgNodeID = (*dfgNode)->getID();
+                int x = refs["Opt2TileXY"][to_string(dfgNodeID)]["x"];
+                int y = refs["Opt2TileXY"][to_string(dfgNodeID)]["y"];
+                refMapRes[*dfgNode] = t_cgra->nodes[y][x];
+                int cgraNodeID = t_cgra->nodes[y][x]->getID();
+                CGRANodeID2Level[cgraNodeID] = refs["Tile2Level"][to_string(cgraNodeID)];
+        }
+
+        /*cout<<"Printing references read from increMapInput.json"<<endl;
+        for(map<DFGNode*, CGRANode*>::iterator iter = refMapRes.begin(); iter != refMapRes.end(); ++iter){
+                cout<<iter->first->getOpcodeName()<<iter->first->getID()<<"->Tile"<<iter->second->getID()<<", level="<<CGRANodeID2Level[iter->second->getID()]<<endl;
+        }
+        cout<<"Done..."<<endl;*/
+
+        return 0;
+}
+
+// generates variables for incremental mapping
+void Mapper::sortAllocTilesByLevel(CGRA* t_cgra){
+        map<int, vector<CGRANode*>> FanIO_CGRANodes;
+        vector<int> FanIOs;
+        int numTiles = 0;
+        for (int i=0; i<t_cgra->getRows(); ++i) {
+                for (int j=0; j<t_cgra->getColumns(); ++j) {
+
+                        // record the number of FanIO for each tile. 
+                        CGRANode* currentCGRANode = t_cgra->nodes[i][j];
+                        if(currentCGRANode->isDisabled()) continue; // only record tiles that have DFGNodes mapped
+                        int FanIO = max(currentCGRANode->getInLinks()->size(), currentCGRANode->getOutLinks()->size());
+                        FanIO_CGRANodes[FanIO].push_back(currentCGRANode);
+
+                        // record FanIO in the list if it appears for the first time
+                        if (find(FanIOs.begin(), FanIOs.end(), FanIO) == FanIOs.end()){
+                                FanIOs.push_back(FanIO);
+                        }
+
+                        numTiles++;
+                }
+        }
+        std::sort(FanIOs.rbegin(), FanIOs.rend()); // sort FanIOs from big to small
```

Comment:
Put comment above the code.

Reply:
Sorting FanIOs from big to small helps automatically form the `Level`
We hope `Level0` should have the list of CGRANodes that have the highest `FanIO`

Reply:
Has been corrected